syntax = "proto3";

package obby.query;

option go_package = "github.com/obby/query-service/proto/generated";

service QueryService {
  // Get recent diffs
  rpc GetRecentDiffs(DiffQuery) returns (stream DiffRecord);

  // Get diffs since timestamp
  rpc GetDiffsSince(SinceQuery) returns (stream DiffRecord);

  // Get file versions
  rpc GetFileVersions(FileQuery) returns (stream FileVersion);

  // Search content
  rpc SearchContent(SearchQuery) returns (stream SearchResult);

  // Get topic files
  rpc GetTopicFiles(TopicQuery) returns (stream FileRecord);

  // Get time analysis
  rpc GetTimeAnalysis(TimeQuery) returns (TimeAnalysisResult);

  // Get activity stats
  rpc GetActivityStats(StatsQuery) returns (ActivityStats);
}

message DiffQuery {
  int32 limit = 1;
}

message SinceQuery {
  int64 timestamp = 1;
  int32 limit = 2;
}

message FileQuery {
  string file_path = 1;
  int32 limit = 2;
}

message SearchQuery {
  string query = 1;
  int32 limit = 2;
}

message TopicQuery {
  string topic = 1;
  int32 limit = 2;
}

message TimeQuery {
  int64 start_timestamp = 1;
  int64 end_timestamp = 2;
}

message StatsQuery {
  int64 start_timestamp = 1;
  int64 end_timestamp = 2;
}

message DiffRecord {
  int64 id = 1;
  string file_path = 2;
  string change_type = 3;
  string diff_content = 4;
  int32 lines_added = 5;
  int32 lines_removed = 6;
  int64 timestamp = 7;
  string content_hash = 8;
  int64 size = 9;
}

message FileVersion {
  int64 id = 1;
  string file_path = 2;
  string content_hash = 3;
  int64 size = 4;
  int64 timestamp = 5;
}

message FileRecord {
  string file_path = 1;
  int64 last_modified = 2;
  int64 size = 3;
}

message SearchResult {
  string file_path = 1;
  string content = 2;
  string highlighted = 3;
  float rank = 4;
}

message TimeAnalysisResult {
  int64 total_files_changed = 1;
  int64 total_lines_added = 2;
  int64 total_lines_removed = 3;
  repeated string top_topics = 4;
  repeated string top_keywords = 5;
}

message ActivityStats {
  int64 files_changed = 1;
  int64 total_changes = 2;
  int64 lines_added = 3;
  int64 lines_removed = 4;
  float avg_changes_per_file = 5;
}

